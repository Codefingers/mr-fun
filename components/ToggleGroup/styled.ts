import { Label } from '@radix-ui/react-label';
import * as ToggleGroupPrimitive from '@radix-ui/react-toggle-group'
import styled from 'styled-components'

/** 
  I know I refer to theme.colors.button in this file for white
  but I saw it used across the app so in the interest of saving time
  and this being a test I'm just going to use it for white. In a real scenario
  I would be looking at renaming the key.
**/

export const ToggleItem = styled(ToggleGroupPrimitive.Item)`
  text-transform: capitalize;
  margin-right: ${({ theme }) => theme.margin[2]};
  border: 4px solid ${({ theme }) => theme.colors.lightGrey};
  background-color: transparent;
  color: ${({ theme }) => theme.colors.button};
  border-radius: ${({ theme }) => theme.borderRadius[1]};

  &:hover { 
    border-color: ${({ theme }) => theme.colors.button};
  }
  
  &[data-state=on] { 
    border-color: ${({ theme }) => theme.colors.violet};
    color: ${({ theme }) => theme.colors.violet};
  }

  &:focus { 
    position: relative; 
    box-shadow: 0 0 0 2px ${({ theme }) => theme.colors.button};
    border-color: rgba(255, 255, 255, .1);
  }

  &:last-of-type {
    margin-right: 0;
  }
`

export const ToggleLabel = styled(Label)`
  font-weight: bold;
  text-transform: uppercase;
`