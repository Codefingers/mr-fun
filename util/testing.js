import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

const Providers = ({ children }) => {
  return (
    <ThemeProvider
      theme={{
        colors: {
          body: '#838fa6',
          heading: '#ebedf2',
          accent: '#BA99FF',
          button: '#FFF',
        },
      }}
    >
      {children}
    </ThemeProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: Providers, ...options });

export * from '@testing-library/react';

// override render method
export { customRender as render };
