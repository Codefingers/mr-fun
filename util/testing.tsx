import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../theme';

export const Providers = ({ children }) => {
  return (
    <ThemeProvider
      theme={theme}
    >
      {children}
    </ThemeProvider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: Providers, ...options });

export * from '@testing-library/react';

// override render method
export { customRender as render };
