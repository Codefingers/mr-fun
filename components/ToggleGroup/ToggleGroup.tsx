import * as ToggleGroupPrimitive from '@radix-ui/react-toggle-group'

import * as Styled from './styled'

interface ToggleGroupProps {
  id: string
  label?: string
  values: Array<string>
  onChange: (key: string) => void
  type?: 'single' | 'multiple'
}

const ToggleGroup = ({ 
  id,
  values,
  label,
  onChange,
  type = 'single' 
}: ToggleGroupProps) => (
  <>
    {label?.length &&  <Styled.ToggleLabel htmlFor={id}>{label}</Styled.ToggleLabel>}
    <ToggleGroupPrimitive.Root type={type} onValueChange={(value) => onChange(value)}>
      {values.map(value => (
        <Styled.ToggleItem key={value} value={value}>{value}</Styled.ToggleItem>
      ))}
    </ToggleGroupPrimitive.Root>
  </>
)

export default ToggleGroup