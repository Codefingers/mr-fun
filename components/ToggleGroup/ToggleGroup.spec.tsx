import userEvent from '@testing-library/user-event'

import { Providers, render, screen } from '../../util/testing'

import ToggleGroup from './ToggleGroup'


const mockOnChange = jest.fn()

const props = {
  id: '__TEST_ID__',
  values: ['_TEST_VALUE_1_', '_TEST_VALUE_2_', '_TEST_VALUE_3_'],
  onChange: mockOnChange
}

const mockLabel = "__TEST_LABEL__"

describe('<ToggleGroup />', () => {
  it('Should render without having an existential crisis', () => {
    const { container } = render(<ToggleGroup {...props} />, { wrapper: Providers })

    expect(screen.getByRole('radio', { name: props.values[0]})).toBeInTheDocument()
    expect(screen.getByRole('radio', { name: props.values[1]})).toBeInTheDocument()
    expect(screen.getByRole('radio', { name: props.values[2]})).toBeInTheDocument()
    expect(screen.queryByRole('label', { name: mockLabel })).not.toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })

  it('Should render with a label', () => {
    render(<ToggleGroup {...props} label={mockLabel} />, { wrapper: Providers })

    expect(screen.getByRole('label', { name: mockLabel })).toBeInTheDocument()
  })

  it('Should fire onChange when clicked', async () => {
    render(<ToggleGroup {...props}  />, { wrapper: Providers })

    const toggleButton = screen.getByRole('radio', { name: props.values[0]})

    // Not sure why I need to await the userEvent here. The test is failing otherwise...
    // That said, with the fact tests are taking so long to run (20s) and
    // how much trouble I had getting them to run in the first place
    // I will leave this as is. Just wanted to flag that I would normally
    // spend more time debugging why this is happening.
    await userEvent.click(toggleButton)

    expect(mockOnChange).toHaveBeenNthCalledWith(1, props.values[0])
  })
})