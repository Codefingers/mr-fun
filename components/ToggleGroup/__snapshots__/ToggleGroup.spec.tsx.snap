// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleGroup /> Should render without having an existential crisis 1`] = `
<div>
  <div
    dir="ltr"
    role="group"
    style="outline: none;"
    tabindex="0"
  >
    <button
      aria-checked="false"
      class="sc-bdVaJa dWTxhh"
      data-radix-collection-item=""
      data-state="off"
      role="radio"
      tabindex="-1"
      type="button"
    >
      _TEST_VALUE_1_
    </button>
    <button
      aria-checked="false"
      class="sc-bdVaJa dWTxhh"
      data-radix-collection-item=""
      data-state="off"
      role="radio"
      tabindex="-1"
      type="button"
    >
      _TEST_VALUE_2_
    </button>
    <button
      aria-checked="false"
      class="sc-bdVaJa dWTxhh"
      data-radix-collection-item=""
      data-state="off"
      role="radio"
      tabindex="-1"
      type="button"
    >
      _TEST_VALUE_3_
    </button>
  </div>
</div>
`;
